scheduler/reconciler/reconcile_node.go

```diff
@@ -419,6 +421,75 @@ func (nr *NodeReconciler) createDeployment(job *structs.Job,
 	nr.DeploymentCurrent.TaskGroups[tg.Name] = dstate
 }
 
+func (nr *NodeReconciler) isDeploymentComplete(groupName string, buckets *NodeReconcileResult) bool {
+	complete := len(buckets.Place)+len(buckets.Migrate) == 0 // && !requiresCanaries   // TODO: additional condition for canaries
+
+	if !complete || nr.DeploymentCurrent == nil {
+		return false
+	}
```

Comment:
I don't think we can fast-path on `(place+migrate) == 0` here. Consider the case where all the allocations were all placed in a previous eval. Once the deploymentwatcher sees that they're healthy, it'll emit a new eval which will result in 0 new placements but `DeploymentCurrent` will be non-nil.

Reply:
Right, but what's what the check below on lines 432-437 is for

Reply:
oooh but then `complete` will be false, I see what you're getting at